Land remaining parts of Torchscript Lazy Tensor backend

wconstab opened the pull request:
Differential Revision: D34772403

wconstab commented:
cc @antoniojkim - I've been working on this for a while but I think it's pretty close to landing

henrytwo commented:
Hey @wconstab, thanks for this PR -- I think this will be very helpful for our efforts. Do you have an idea of when this branch will be stable/ready to land in master? I noticed that there are some differences in the codegen compared to what we have in `lazy_tensor_staging`

wconstab commented on CMakeLists.txt:
```diff
@@ -337,6 +337,9 @@ cmake_dependent_option(USE_CCACHE "Attempt using CCache to wrap the compilation"
 option(WERROR "Build with -Werror supported by the compiler" OFF)
 option(USE_COREML_DELEGATE "Use the CoreML backend through delegate APIs" OFF)
 option(USE_PER_OPERATOR_HEADERS "Whether ATen should generate separate headers for each operator" ON)
+cmake_dependent_option(
```
@malfet Any chance you can spot what I'm doing wrong here? If you look at failures on this PR's CI, I am observing from cmake.summary that 
INTERN_BUILD_MOBILE: ON
BUILD_LAZY_TS_BACKEND: ON
and I can't figure out why my flag isn't getting turned off under these builds.

antoniojkim commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -235,12 +240,13 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
-            "lazy_tensor_core/csrc/ts_backend/aten_eager_fallback.h",
-            f"{output_dir}/{backend_key}NativeFunctions.h",
-            f"{output_dir}/{backend_key}LazyIr.h",
+            f"torch/csrc/lazy/generated/{backend_key}NativeFunctions.h",
+            "torch/csrc/lazy/generated/LazyIr.h",
```
@wconstab Removing this `output_dir` argument effectively forces all LTC backends to use only the generated `LazyIr.h` and `LazyNativeFunctions.h` from the TS backend. This needs to be changes to accept a variable `output_dir` again and set the `output_dir` for the default case to be`torch/csrc/lazy/generated`

wconstab replied:
Good catch,  I'm not sure why I did that, maybe there was a reason, but I will try to put it back as an argument.

wconstab commented:
> Hey @wconstab, thanks for this PR -- I think this will be very helpful for our efforts. Do you have an idea of when this branch will be stable/ready to land in master (e.g. in a few days, or a month)? I noticed that there are some differences in the codegen compared to what we have in `lazy_tensor_staging`, so I want to make sure it's stable before using it

@henrytwo ASAP, in a matter of days hopefully.  I think I just have one minor cmake setting issue now, but one never knows how many more attempts that will take.

henrytwo commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -222,9 +223,13 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
 
     # Generate Dispatcher registrations which hook up the nativefunctions
+    internal_headers = [
+        "torch/csrc/lazy/generated/LazyNativeFunctions.h",
```
@wconstab I think a similar change to what was discussed below with the `output_dir` needs to be applied here too, based on my testing

henrytwo commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -251,7 +251,7 @@ def this_shape(i: int) -> str:
         TORCH_INTERNAL_ASSERT(shapes.size() == {returns_length});"""
 
-        node_str = f"""auto node = torch::lazy::MakeNode<ir::ops::{schema.node_name}>({node_ctor_input_str},
+        node_str = f"""auto node = torch::lazy::MakeNode<{schema.node_name}>({node_ctor_input_str},
```
Why was this changed to exclude the namespace? Should nodes no longer be under the `ir::ops` namespace?

wconstab replied:
yea, generally we've simplified the namespaces when landing to master.  The ir::ops. namespace was inherited from the torch/xla project.  We use a flatter namespace, almost everything is just under torch::lazy:: now.

wconstab commented:
Moving to land with several failing CUDA and ASAN unit tests (disabled) and fix separately, because they are also happening on lazy_tensor_staging branch (which only runs CPU mode in CI and no ASAN).

Issues tracked here:
https://github.com/pytorch/pytorch/issues/74519

wconstab committed:
Make lazy codegen honor per-operator-headers flag

Summary:
- per-operator-headers is a strict build mode where compulation units aren't allowed
to depend on bulk headers like ATen/Functions.h, but must instead depend only on the
specific operator headers used.  (In other configurations, the reverse is required).

Differential Revision: D35002666

fbshipit-source-id: a1747cf994fa1dc55095dceebe67cc819dee03d5

wconstab committed:
Land remaining parts of Torchscript Lazy Tensor backend (#74111)

Summary:
Also enables bazel build to run lazy codegen.  Bazel (oss) build feeds off the same filelists as cmake/buck (build_variables.bzl), so enabling it is easier than keeping it disabled.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/74111

Test Plan: Run CI and verify test_lazy_ops is running via OSS cmake builds

Differential Revision: D34772403

fbshipit-source-id: 55a3db507ef9f5bfbb13e8e4692984b2ad6fa065

bdhirsh commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -613,14 +614,13 @@ void LazyBackendsAutogradOverridesAutogradKernel(DispatchKey key) {
 
+// no longer test ::Lazy key here
+// since it is now registered to TS backend in-tree and thus behaves differently,
+// does not throw the expected 'could not run..' messages
```
I was trying to think of why your changes would affect these tests, since they all use a custom operator (`_test::dummy`). And... it's probably because we now register a CPU fallback kernel to the Lazy key in core? It would probably be better if we just arranged for this test to get built without any of the dispatcher registrations running :p.

Either way, removing seems totally fine.

wconstab replied:
yes, the fallback affects these.

actually, there is another PR out (https://github.com/pytorch/pytorch/pull/73973) which I wanted you to take a look at, and I plan to land after this.  Once that's in, i think we could more easily run tests like this, however, it also doesn't seem like that important of a test to me given the other coverage we'll be adding

bdhirsh commented on caffe2/CMakeLists.txt:
```diff
@@ -350,18 +350,25 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       "${TORCH_SRC_DIR}/csrc/autograd/generated/ADInplaceOrViewType_0.cpp"
       "${TORCH_SRC_DIR}/csrc/autograd/generated/ADInplaceOrViewType_1.cpp"
     )
+    if(BUILD_LAZY_TS_BACKEND)
```
Mostly just my check for understanding - we're going to unconditionally run the LTC codegen and emit the C++ files, right? And then actually building them looks like it will get guarded by this build flag. (I think this seems fine - it's probably not worth the work to make the codegen conditionally emit LTC code if we don't end up building it).

wconstab replied:
Right. The codegen runs always, which we could try to avoid i suppose.

bdhirsh replied:
(nah seems fine to me)

bdhirsh commented on test/cpp/lazy/test_tensor_impl.cpp:
```diff
@@ -6,12 +6,14 @@
 
-// TODO(alanwaketan): Update the following unit tests once the TorchScript backend is merged.
+#ifdef FBCODE_CAFFE2
+// Lazy Tensor is disabled in FBCODE until addressing non-virtual methods (e.g. sizes) in TensorImpl
```
If this ends up being important, the sizes thing is fixable - @swolchok recently landed something to fix a similar issue for NestedTensor: https://github.com/pytorch/pytorch/pull/73817

desertfire commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -43,7 +43,9 @@ def gen_registration_headers(
             "#include <ATen/ops/empty.h>",
-            "#include <ATen/ops/empty_strided.h>"]
+            "#include <ATen/ops/empty_strided.h>",
+            "#include <ATen/ops/_copy_from_and_resize.h>",
```
I don't fully understand how `per_operator_headers` works, so just curious why we need to add the extra headers here?

bdhirsh replied:
tldr: we used to get away with `#include <ATen/Functions.h>`, which would include every tensor operator (easy for development, but bloats all of our C++ files). The per_operator_headers stuff basically generates a new header file for every aten operator, and enforces that you only include the headers for the operators that you actually need.

The codegen uses the metadata in `native_functions.yaml` to figure out which #includes to stamp out, but there are a few ops that are hardcoded directly into the `RegisterDispatchKey.cpp` template file (that we need to hardcode here, because the codegen would know to include them just based on the yaml).

My guess for why we needed that extra `_copy_from_and_resize` op is because the LTC codegen uses it directly (to handle inplace/out= ops)

bdhirsh commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -241,15 +241,25 @@ def gen_dispatcher_registrations(
         dispatch_key: DispatchKey,
-        selector: 'SelectiveBuilder') -> None:
+        selector: 'SelectiveBuilder',
+        build_in_tree: bool = False,
```
Why do we need the flag in `gen_backend_stubs.py` (I thought it wasn't used by LTC?)

Also, I think a light comment here on why we have `build_in_tree` would be nice. Basically, `gen_backend_stubs.py` was originally meant as a codegen pipeline for external backends that don't have the ability to take advantage of `native_functions.yaml` and our core codegen pipeline. And now that LTC is a part of core, it may or may not get built in-tree (depending on whether the backend that hooks into LTC is in-tree or not).

bdhirsh replied:
ohh it's just because `gen_lazy_tensor.py` calls into the function in `gen_backend_stubs.py` (so if you're adding a new flag for the LTC codegen, the backend_stubs function needs to know about it). I guess that's fine, since the `gen_backend_stubs.py` entrypoint isn't affected - just the helper functions it uses. (Maybe just include in the comment here that it's only actually needed/used by LTC codegen)

wconstab replied:
I added a comment locally, but if my CI is clean this time I'm gonna land the comment in a separate diff haha

bdhirsh reviewed (approved):
I looked at the codegen, and skimmed the build system changes (but didn't look in detail at the LTC C++ code changes). Looks good!

wconstab commented on torch/csrc/lazy/ts_backend/ts_backend_impl.cpp:
```diff
@@ -0,0 +1,233 @@
+ public:
+  TSBackendImpl() : default_device_type_(at::kCPU) {
+    // TODO(whc) unify how all our flags are set and parsed as envs
+    static bool env_use_cuda = std::getenv("LTC_TS_CUDA") != nullptr;
```
yea, it is not ideal.  I copied the way that LTC_.._dynamic shapes mode is set (see ir.cpp)

I think we should change this, but propose to unify them all in a later PR and have one system that lets us override any of the 'FLAGS_' variables using envs with a consistent naming.

wconstab commented on tools/setup_helpers/generate_code.py:
```diff
@@ -184,24 +189,27 @@ def main() -> None:
         aten_path = os.path.dirname(os.path.dirname(options.native_functions_path))
         ts_backend_yaml = os.path.join(aten_path, 'native/ts_native_functions.yaml')
-
+        ts_native_functions = "torch/csrc/lazy/ts_backend/ts_native_functions.cpp"
```
> I'll leave it up to you, but the idea with the impl_path argument ...

Wait, i'm confused- It sounds like you think I did not use impl_path and am not opting into extra sanity checking fo ts_native_functions.cpp

I actually think I am opting into the sanity checking and the whole reason I hardcoded the ts_native_functions path here is to enable that.  If I missed something lmk?

bdhirsh replied:
oh nope I just phrased that poorly - you're using it, I just wanted to let you know that it's actually meant to be optional (and gives you some extra error checking) / let you know the history.

desertfire reviewed (approved):
LTC C++ part looks good to me.
